gui/README.md

```diff
@@ -15,6 +15,23 @@ The layout engine does a top-down pass through the component tree in order to de
 
 Layout is controlled using predefined attributes, such as `width`, `height`, `x-align`, `y-align`, `spacing` or `padding`.
 
+### Layout algorithm
+
+To calculate dimensions (width/height):
+- elements with **fixed** sizes (e.g. `123px`) are laid out, in the order they are encountered.
+- then **percentages** (such as `10%`) are computed, based on the total size of the parent container.
+- then we handle elements with `inner`, which indicates that the size of the component depends on the size of the child.
+- then `@` rules (e.g. `100@`) are applied, which divide up the remaining free space.
```

Comment:
`inner` elements will actually need to be calculated before `%` elements.

Think of `px` and `inner` as elements with outwards-pushing sizes. Their demanded size is set within (either the literal `px` value or by recursing into the `inner` element and solving for a pixel value). Then padding and spacing are added outside of them (padding grows the actual size, spacing just pushes the x/y alignment within its container and takes up remaining space).

After these two are calculated (`px` then `inner`), the container has been pushed to expand by these types of children. Its size is now known, so it can go push inwards on `%` and `@` elements.

Think of `%` and `@` as elements with sizes that are pushed inwards by their container. The container's size is now known, and these two types of elements have their dimensions based on that known size. `%` elements are calculated and that means all the dimensions are known except `@` elements. Add up all the known sizes and subtract that sum from the container's dimension to get the remaining unfilled pixels. Now the `@` elements are finally given a size equal to their percentage of the remainder. (`50@` = 50% of the remainder, so a `100px` remainder becomes an absolute `50px` size for that element.)

So in summary, the order goes like this:
1. Immediately count all the `px` values, these are done.
2. Recursively calculate `inner` sizes to get their pixel values, now all outwards-pushing elements are done.
3. If the container was `inner`, it now has a known size: `max(outwardsPusher + itsPadding + itsSpacing, ...)` pixels
4. Now `%` values are `percentage / 100 * containerSize` pixels
5. Finally `@` values are `atSignPercentage / 100 * (containerSize - (knownElementSizes + theirPadding + theirSpacing))` pixels

Reply:
Thanks for the explanation, I've updated the `README.md` for GUI with a summary.